src/compositemark/errorbar.ts

```diff
@@ -1,42 +1,395 @@
+    continuousAxis = 'x';
+    continuousAxisChannelDef = encoding.x as FieldDef<string>; // Safe to cast because if x is not continuous fielddef, the orient would not be horizontal.
+  }
+
+  if (continuousAxisChannelDef && continuousAxisChannelDef.aggregate) {
+    const {aggregate, ...continuousAxisWithoutAggregate} = continuousAxisChannelDef;
+    if (aggregate !== ERRORBAR) {
+      log.warn(`Continuous axis should not have customized aggregation function ${aggregate}`);
+    }
+    continuousAxisChannelDef = continuousAxisWithoutAggregate;
+  }
+
+  return {
+    continuousAxisChannelDef,
+    continuousAxis
+  };
+}
+
+function isExtentCI(extent: ErrorBarExtent): extent is 'ci' {
+  return extent === 'ci';
+}
+
+function isExtentIqr(extent: ErrorBarExtent): extent is 'iqr' {
+  return extent === 'iqr';
+}
+
+function errorBarSpecialExtents(extent: 'ci' | 'iqr', fieldName: string): AggregatedFieldDef[] {
+  return [
+    {
+      op: (extent === 'ci') ? 'ci0' : 'q1',
+      field: fieldName,
+      as: 'lower_rule_' + fieldName
+    },
+    {
+      op: (extent === 'ci') ? 'ci1' : 'q3',
+      field: fieldName,
+      as: 'upper_rule_' + fieldName
+    }
+  ];
+}
+
+function errorBarParams(spec: GenericUnitSpec<Encoding<string>, ErrorBar | ErrorBarDef>, orient: Orient, center: ErrorBarCenter, extent: ErrorBarExtent) {
+
+  const {continuousAxisChannelDef, continuousAxis} = errorBarContinousAxis(spec, orient);
+  const encoding = spec.encoding;
+
+  const continuousFieldName: string = continuousAxisChannelDef.field;
+
+  let aggregate: AggregatedFieldDef[] = [
+    {
+      op: center,
+      field: continuousFieldName,
+      as: center + '_' + continuousFieldName
+    }
+  ];
+
+  let postAggregateCalculates: CalculateTransform[] = [];
+
+  if (isExtentCI(extent)) {
+    aggregate = aggregate.concat(errorBarSpecialExtents('ci', continuousFieldName));
+  } else if (isExtentIqr(extent)) {
+    aggregate = aggregate.concat(errorBarSpecialExtents('iqr', continuousFieldName));
+  } else {
+    aggregate.push({
+      op: extent,
+      field: continuousFieldName,
+      as: 'extent_' + continuousFieldName
+    });
+
+    postAggregateCalculates = [{
+        calculate: `datum.${center}_${continuousFieldName} + datum.extent_${continuousFieldName}`,
+        as: 'upper_rule_' + continuousFieldName
+      },
       {
-        mark: 'rule',
-        encoding: encodingWithoutSize
-      },{ // Lower tick
-        mark: 'tick',
-        encoding: encodingWithoutX2Y2
-      }, { // Upper tick
-        mark: 'tick',
-        encoding: encoding.x2 ? {
-          x: encoding.x2,
-          y: encoding.y,
-          ...encodingWithoutX_X2_Y_Y2
-        } : {
-          x: encoding.x,
-          y: encoding.y2,
-          ...encodingWithoutX_X2_Y_Y2
+        calculate: `datum.${center}_${continuousFieldName} - datum.extent_${continuousFieldName}`,
+        as: 'lower_rule_' + continuousFieldName
+    }];
+  }
+
+
+  const groupby: string[] = [];
```

Comment:
This part for producing `groupBy, bins, timeUnit`  seems exactly the same as boxplot?  Extract a method to `common.ts`.

Reply:
Yes. Sorry didn’t call catch that